test/common/config/utility_test.cc

```diff
@@ -283,8 +253,65 @@ TEST(UtilityTest, FactoryForApiConfigSource) {
     api_config_source.add_grpc_services()->mutable_envoy_grpc()->set_cluster_name("foo");
     EXPECT_CALL(async_client_manager,
                 factoryForGrpcService(ProtoEq(api_config_source.grpc_services(0)), Ref(scope), _));
-    Utility::factoryForApiConfigSource(async_client_manager, api_config_source, scope);
+    Utility::factoryForGrpcApiConfigSource(async_client_manager, api_config_source, scope);
+  }
+}
+
+class CheckApiConfigSourceSubscriptionBackingClusterTest
+    : public testing::TestWithParam<::envoy::api::v2::core::ApiConfigSource_ApiType> {};
+
+INSTANTIATE_TEST_CASE_P(CheckApiConfigSourceSubscriptionBackingClusterTest,
+                        CheckApiConfigSourceSubscriptionBackingClusterTest,
+                        ::testing::Values(envoy::api::v2::core::ApiConfigSource::REST_LEGACY,
+                                          envoy::api::v2::core::ApiConfigSource::REST,
+                                          envoy::api::v2::core::ApiConfigSource::GRPC));
+
+TEST_P(CheckApiConfigSourceSubscriptionBackingClusterTest, NonExistentCluster) {
+  envoy::api::v2::core::ConfigSource config;
+  auto* api_config_source = config.mutable_api_config_source();
+  Upstream::ClusterManager::ClusterInfoMap cluster_map;
+  api_config_source->set_api_type(GetParam());
+  if (GetParam() == envoy::api::v2::core::ApiConfigSource::GRPC) {
+    // Test case for gRPC API configs with cluster names set instead of gRPC services.
+    // Even though this will be deprecated, we should still throw on statically defined non-EDS
+    // clusters.
+    api_config_source->add_grpc_services();
   }
+  api_config_source->add_cluster_names("foo_cluster");
```

Comment:
You'll probably want another standalone test for the case where cluster_names.size() != 0 for gRPC but grpc_services.size() == 0.

Reply:
Reworked this in d62249f.